src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+#endif
+    }
+
+    public SourceGeneratedDocumentState? FrozenSourceGeneratedDocumentState => _frozenSourceGeneratedDocumentState;
+
+    private SolutionCompilationState Branch(
+        SolutionState newSolutionState,
+        ImmutableDictionary<ProjectId, ICompilationTracker>? projectIdToTrackerMap = null,
+        Optional<SourceGeneratedDocumentState?> frozenSourceGeneratedDocument = default)
+    {
+        projectIdToTrackerMap ??= _projectIdToTrackerMap;
+        var newFrozenSourceGeneratedDocumentState = frozenSourceGeneratedDocument.HasValue ? frozenSourceGeneratedDocument.Value : _frozenSourceGeneratedDocumentState;
+
+        if (newSolutionState == this.Solution &&
+            projectIdToTrackerMap == _projectIdToTrackerMap &&
+            newFrozenSourceGeneratedDocumentState == _frozenSourceGeneratedDocumentState)
+        {
+            return this;
+        }
+
+        return new SolutionCompilationState(
+            newSolutionState,
+            PartialSemanticsEnabled,
+            projectIdToTrackerMap,
+            newFrozenSourceGeneratedDocumentState);
+    }
+
+    /// <inheritdoc cref="SolutionState.ForkProject"/>
+    private SolutionCompilationState ForkProject(
+        (SolutionState newSolutionState, ProjectState oldProjectState, ProjectState newProjectState) stateTuple,
+        CompilationAndGeneratorDriverTranslationAction? translate,
+        //ProjectDependencyGraph? newDependencyGraph = null,
+        bool forkTracker)
+    {
+        return ForkProject(
+            stateTuple.newSolutionState,
+            stateTuple.newProjectState,
+            translate,
+            forkTracker);
+    }
+
+    /// <inheritdoc cref="SolutionState.ForkProject"/>
+    private SolutionCompilationState ForkProject(
+        SolutionState newSolutionState,
+        ProjectState newProjectState,
+        CompilationAndGeneratorDriverTranslationAction? translate,
+        //ProjectDependencyGraph? newDependencyGraph = null,
+        bool forkTracker)
+    {
+        // If the spsolution didn't actually change, there's no need to change us.
+        if (newSolutionState == this.Solution)
+            return this;
+
+        return ForceForkProject(newSolutionState, newProjectState, translate, forkTracker);
+    }
+
+    /// <summary>
+    /// Same as <see cref="ForkProject(SolutionState, ProjectState, CompilationAndGeneratorDriverTranslationAction?,
+    /// bool)"/> except that it will still fork even if <paramref name="newSolutionState"/> is unchanged from
+    /// <see cref="Solution"/>.
+    /// </summary>
+    private SolutionCompilationState ForceForkProject(
+        SolutionState newSolutionState,
+        ProjectState newProjectState,
+        CompilationAndGeneratorDriverTranslationAction? translate,
+        bool forkTracker)
+    {
+        var projectId = newProjectState.Id;
+
+        var newDependencyGraph = newSolutionState.GetProjectDependencyGraph();
```

Comment:
in solutionState.ForkProject, we create this dep graph and pass it into the new solution state.  It's a O(1) op to retrieve this as it's not actually computing anything (we should really make that method a property to indicate that)